### ISSUE TITLE
Port build system to CMake

### ISSUE THREAD

**Issue by TheMarex** [valhalla/valhalla#1251]
The GNU build system is more than antique by now and should be replaced by something that supports cross-platform out-of-tree builds. The industry standard for modern C++ code bases is to use CMake which is supported on all major and minor operating systems, and actively maintained.

Looking at the code base there is already an attempt for porting to CMake, but it seems incomplete? It looks like it uses some CMake anti-patterns like hard coding file paths: [here](https://github.com/valhalla/valhalla/blob/master/CMakeLists.txt#L194-L327). CMake has a good mechanism for this that basically allows you to use regular expressions for this ([example](https://github.com/Project-OSRM/osrm-backend/blob/master/CMakeLists.txt#L139-L149)).

The port to CMake is especially important since it supports good toolchain for building on ARM, which is important for offline.

**Comment by danpat** [comment]
Another plus is that this would avoid some of the `libtool` wrapper logic which adds extra hoops when debugging tests/tools.

**Comment by kevinkreiser** [comment]
I agree that removing the libtiol wrapper is good but I find testing and building in general is more of a hassle with cmake.  Like the test targets don't know to build the library target they depend on!? Maybe someone has made a better test framework for cmake that we could use.

Also I don't know what challenges cmake has in terms of making debian packages. Something we should keep in mind

**Comment by kevinkreiser** [comment]
I neglected to mention that I do think @TheMarex is right about this being a good idea for cross platform and out of tree builds. I just want to make sure that we dont lose any of the things that we currently have like:

* conditional builds via --enable-services, --enable-python-bindings,  --enable-data-tools, --enable-coverage
* easy interoperability with debian packages
* testing locales (at least on linux)

**Comment by mloskot** [comment]
@TheMarex As the author of the original `CMakeLists.txt` I will clarify couple of your points:
> Looking at the code base there is already an attempt for porting to CMake, but it seems incomplete?

Yes, it is incomplete. The status of the CMake configuration is explained in the header comment of the `CMakeLists.txt`.

> It looks like it uses some CMake anti-patterns like hard coding file paths:

One could argue that the `GLOB`-ing technique used by OSRM is the actual anti-pattern as per the widely-accepted "Modern CMake" recommendation: [Do not use file(GLOB in projects](https://www.youtube.com/watch?v=bsXLMQ6WgIk&t=15m29s).

However, I deliberately submitted an old-style `CMakeLists.txt` as a minimal as possible single file configuration to make it as non-intrusive to existing official build configuration (and source tree) as possible. That, I assumed, would also increase chances for accepting my proposal to also support CMake.
The old-style `CMakeListst.txt` are also often easier to understand by those not experienced with CMake - it was also one of key factors that shaped my proposal.

Finally, since Valhalla team seems to be excited about switching to CMake, it is a good idea to refactor the minimal single-file CMake configuration to proper one following the modern CMake approach.
I'd like to thank you @TheMarex and @oxidase for taking the lead on this initiative.

**Comment by mloskot** [comment]
@oxidase & @TheMarex Thanks for the overhaul, much appreciated.

**Comment by mloskot** [comment]
This may be relevant for Valhalla setup (and perhaps OSRM too)

CMake 3.12 has got improved globbing, https://cmake.org/cmake/help/v3.12/release/3.12.html that 

> The file(GLOB) and file(GLOB_RECURSE) commands learned a new flag CONFIGURE_DEPENDS
> which enables expression of build system dependency on globbed directory’s contents.

This clears one of the modern CMake points against `GLOB`-ing for source files, so `cmake --build .` can now check if any `GLOB`-ed files have been added/removed.
